src/blueprint/from_config.rs

```diff
@@ -207,7 +207,100 @@ fn to_fields(type_of: &config::Type, config: &Config) -> Valid<Vec<blueprint::Fi
+  }
+
+  let head = parts[0].as_str();
+  let tail = parts[1].as_str();
+
+  match head {
+    "value" => {
+      if let Some(val) = get_value(tail) {
+        if val.of_type.is_nullable() {
+          return Valid::fail(format!("value '{tail}' is a nullable type"));
+        }
+      } else {
+        return Valid::fail(format!("no value '{tail}' found"));
+      }
+    }
+    // TODO this should probably error out on the `ListType` match arm since it
+    // doesn't make sense to have a list in the template URL
+    "args" => {
+      // XXX this is a linear search but it's cost is less than that of
+      // constructing a HashMap since we'd have 3-4 arguments at max in
+      // most cases
+      if let Some(arg) = args.iter().find(|arg| arg.name == tail) {
+        if arg.default_value.is_none() && arg.of_type.is_nullable() {
+          return Valid::fail(format!("argument '{tail}' is a nullable type"));
+        }
+      } else {
+        return Valid::fail(format!("no argument '{tail}' found"));
+      }
+    }
+    "headers" | "vars" => {
+      // "headers" refers to the header values known at runtime, which we can't
+      // validate here
+
+      // "vars" refer to the server's configuration variables, set up in
+      // `config/server.rs` and they all have fallbacks to default values, so
+      // we don't need any checks here
+    }
+    _ => {
+      return Valid::fail(format!("unknown template directive '{head}'"));
+    }
+  }
+
+  Valid::Ok(())
+}
+
+fn validate_fields(fields: &[FieldDefinition]) -> Valid<()> {
+  let mut validation_map = HashMap::<&str, &FieldDefinition>::new();
+  validation_map.reserve(fields.len());
+
+  for field in fields {
+    if validation_map.insert(&field.name, field).is_some() {
+      // Fields are already de-duplicated for us beforehand
+      panic!("Field '{}' shouldn't be already present!", field.name);
+    }
+  }
+
+  for field in fields {
+    // XXX we could use `Mustache`'s `render` method with a mock
+    // struct implementing the `PathString` trait encapsulating `validation_map`
+    // but `render` simply falls back to the default value for a given
+    // type if it doesn't exist, so we wouldn't be able to get enough
+    // context from that method alone
+    // So we must duplicate some of that logic here :(
+    if let Some(Expression::Unsafe(Operation::Endpoint(req_template, _, _))) = &field.resolver {
+      for parts in req_template.root_url.all_parts() {
+        validate_mustache_parts(parts, &field.args, |k| validation_map.get(k).copied()).trace(&field.name)?;
+      }
+
+      for query in &req_template.query {
+        let (name, mustache) = query;
+
+        for parts in mustache.all_parts() {
+          validate_mustache_parts(parts, &field.args, |k| validation_map.get(k).copied())
+            .trace(name)
+            .trace(&field.name)?;
+        }
+      }
+    }
+  }
+
+  Valid::Ok(())
 }
 
 fn to_field(
```

Comment:
I think we can clean this up 
- Add a function that gets a type at a particular path 
  ```rust
  fn infer_type(&self, path: Vec<String>) -> Option<(&str, &Type)> {  

  }
  ```
  This function will just go thru the config and try to infer the type at a particular path.

- Then use the `segments` function from mustache to validate if that path exists.

Reply:
- Also note there are scenarios where values are optional, so it shouldn't fail if the path doesn't exist.
- We need to make sure that the value at that particular path is a "string" or some "Scalar" type or else validation should fail.

Reply:
I think there's already [`process_path`](https://github.com/tailcallhq/tailcall/blob/0061a2419211dbd4226a373bfda41ef57fe1f5a1/src/blueprint/from_config.rs#L534) that can infer types at a given path, given the `Config` struct

Now that is fine, but the actual traversal logic where we go through Mustache directives like `args`, `value`, etc. will have to remain in `validate_fields` only

So the logic would look something like so

```rs
fn validate_fields(...) {
    type_at_current_traversal = ...; // say `Post`
    match head {
        "value" => process_path([type_at_current_traversal, tail]), // say `[Post, userId]`
        "args" => process_path([??, tail]) // not sure how to get the "path" to an argument
}
```

Does this sound good?

Reply:
Actually wait, I forgot that we already know the types of the fields here, so we can just check the `of_type` field for scalar values, just like how the nullable check is being performed right now..